sw/device/lib/crypto/drivers/keymgr.c

```diff
@@ -0,0 +1,257 @@
+  do {
+    reg = abs_mmio_read32(kBaseAddr + KEYMGR_OP_STATUS_REG_OFFSET);
+    status = bitfield_field32_read(reg, KEYMGR_OP_STATUS_STATUS_FIELD);
+  } while (status == KEYMGR_OP_STATUS_STATUS_VALUE_WIP);
+
+  // Clear OP_STATUS by writing back the value we read.
+  abs_mmio_write32(kBaseAddr + KEYMGR_OP_STATUS_REG_OFFSET, reg);
+
+  // Check if the key manager reported errors. If it is already idle or
+  // completed an operation successfully, return an OK status.
+  switch (status) {
+    case KEYMGR_OP_STATUS_STATUS_VALUE_IDLE:
+      return OTCRYPTO_OK;
+    case KEYMGR_OP_STATUS_STATUS_VALUE_DONE_SUCCESS:
+      return OTCRYPTO_OK;
+    case KEYMGR_OP_STATUS_STATUS_VALUE_DONE_ERROR: {
+      // Clear the ERR_CODE register before returning.
+      uint32_t err_code =
+          abs_mmio_read32(kBaseAddr + KEYMGR_ERR_CODE_REG_OFFSET);
+      abs_mmio_write32(kBaseAddr + KEYMGR_ERR_CODE_REG_OFFSET, err_code);
+      return OTCRYPTO_RECOV_ERR;
+    }
+    case KEYMGR_OP_STATUS_STATUS_VALUE_WIP:
+      // Should be unreachable.
+      HARDENED_TRAP();
+      return OTCRYPTO_FATAL_ERR;
+    default:
+      // Should be unreachable.
+      HARDENED_TRAP();
+      return OTCRYPTO_FATAL_ERR;
+  }
+
+  // Should be unreachable.
+  HARDENED_TRAP();
+  return OTCRYPTO_FATAL_ERR;
+}
+
+/**
+ * Set the control register of the key manager.
+ *
+ * The CDI select bit is always set to false for this driver.
+ *
+ * @param dest (NONE, AES, OTBN, or KMAC)
+ * @param operation (GENERATE_SW or GENERATE_HW)
+ */
+#define WRITE_CTRL(dest, operation)                                            \
+  do {                                                                         \
+    uint32_t ctrl =                                                            \
+        bitfield_field32_write(0, KEYMGR_CONTROL_SHADOWED_DEST_SEL_FIELD,      \
+                               KEYMGR_CONTROL_SHADOWED_DEST_SEL_VALUE_##dest); \
+    ctrl = bitfield_bit32_write(ctrl, KEYMGR_CONTROL_SHADOWED_CDI_SEL_BIT,     \
+                                false);                                        \
+    ctrl = bitfield_field32_write(                                             \
+        ctrl, KEYMGR_CONTROL_SHADOWED_OPERATION_FIELD,                         \
+        KEYMGR_CONTROL_SHADOWED_OPERATION_VALUE_##operation##_OUTPUT);         \
+    abs_mmio_write32_shadowed(kBaseAddr + KEYMGR_CONTROL_SHADOWED_REG_OFFSET,  \
+                              ctrl);                                           \
+  } while (false);
+
+status_t keymgr_generate_key_sw(keymgr_diversification_t diversification,
+                                keymgr_output_t *key) {
+  HARDENED_TRY(keymgr_is_idle());
+
+  // Set the control register to generate a software-visible key.
+  WRITE_CTRL(NONE, GENERATE_SW);
+
+  // Check that the version does not exceed the max owner key version.
+  uint32_t max_version =
+      abs_mmio_read32(kBaseAddr + KEYMGR_MAX_OWNER_KEY_VER_SHADOWED_REG_OFFSET);
```

Comment:
This assumes that the library will only be used when the `keymgr` has reached the `OWNER` stage. The `ROM_EXT` may have some use cases for generating keys, although it can probably limit use to sideloaded keys.

I think if you want to keep this code as is, it probably makes sense to check the current state of the `keymgr`. Otherwise I would recommend to use the current internal state to select the appropriate `max ver` register to compare against.

Side note: the hardware is expected to return an error if the diversification version is greater than the current `max_version`. This error may manifest itself as `INVALID_OP` with `DEBUG == INVALID_KEY_VERSION`

Reply:
My understanding was that the cryptolib drivers were exclusively for the OS-visible crypto library, and that `ROM_EXT` would use the drivers under `silicon_creator`. Or is it the case that the cryptolib drivers are for `ROM_EXT` and cryptolib, and the `silicon_creator` drivers are for `ROM` only? I suppose that makes sense from a binary separation standpoint.

As for why I checked the result in software here even though it would result in a hardware error: this is mostly just to help the caller by returning `OTCRYPTO_BAD_ARGS` instead of `OTCRYPTO_RECOV_ERR`, since I thought that was a more appropriate message for this case. Not absolutely necessary to do this, though, especially if it would complicate the code.

Reply:
Maybe for now I'll just remove the check; the caller will get an error anyway and this way we don't introduce the possibility of an error in a case that shouldn't actually cause one.

Reply:
I think it is ok to keep, but it would then make sense to check that the `keymgr` is in the expected state. 

I think in general it makes sense for `ROM_EXT` to use cryptolib, but it would be interesting to figure out first how we are planning to provide crypto functionality that overlaps both the Silicon Creator and Silicon Owner (e.g. attestation signatures). Other than attestation, I foresee the need for sideloading keys into KMAC for integrity calculations in `ROM_EXT`; and, entropy complex initialization. I am not sure if I am missing anything else.